File: cs-cli/src/bin/view_atm_iv.rs
// Simple utility to view ATM IV parquet files

use polars::prelude::*;
use std::env;
use std::process;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 {
        eprintln!("Usage: {} <parquet_file>", args[0]);
        process::exit(1);
    }

    let file_path = &args[1];

    // Read parquet file
    let df = LazyFrame::scan_parquet(file_path, Default::default())?
        .collect()?;

    println!("{}", "=".repeat(100));
    println!("ATM IV Time Series: {}", file_path);
    println!("{}", "=".repeat(100));
    println!();

    println!("Shape: {} rows × {} columns", df.height(), df.width());
    println!();

    // Convert date column from days since epoch to readable format
    let df_display = df.clone().lazy()
        .with_column(
            col("date")
                .cast(DataType::Date)
                .alias("date")
        )
        .collect()?;

    // Display data
    println!("{}", df_display);
    println!();

    // Calculate statistics
    println!("{}", "=".repeat(100));
    println!("Summary Statistics:");
    println!("{}", "=".repeat(100));

    let stats_cols = vec!["atm_iv_30d", "atm_iv_60d", "atm_iv_90d", "term_spread_30_60", "term_spread_30_90"];

    for col_name in stats_cols {
        if let Ok(col) = df.column(col_name) {
            if let Ok(stats) = col.f64() {
                let valid: Vec<f64> = stats.into_iter().filter_map(|x| x).collect();
                if !valid.is_empty() {
                    let min = valid.iter().cloned().fold(f64::INFINITY, f64::min);
                    let max = valid.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
                    let mean = valid.iter().sum::<f64>() / valid.len() as f64;
                    println!("{:20} count={:3}, min={:.4}, max={:.4}, mean={:.4}",
                        col_name, valid.len(), min, max, mean);
                }
            }
        }
    }
    println!();

    // Detect earnings signals
    println!("{}", "=".repeat(100));
    println!("Potential Earnings Signals:");
    println!("{}", "=".repeat(100));

    if let (Ok(dates), Ok(iv_30d), Ok(spreads)) = (
        df.column("date"),
        df.column("atm_iv_30d"),
        df.column("term_spread_30_60"),
    ) {
        let dates_i32 = dates.i32()?;
        let iv_30d = iv_30d.f64()?;
        let spreads = spreads.f64()?;

        let mut found_signals = false;

        for i in 1..df.height() {
            let date_days = dates_i32.get(i);
            let iv_curr = iv_30d.get(i);
            let iv_prev = iv_30d.get(i - 1);
            let spread = spreads.get(i);

            let mut signals = Vec::new();

            // IV change detection
            if let (Some(curr), Some(prev)) = (iv_curr, iv_prev) {
                if prev > 0.0 {
                    let change_pct = (curr - prev) / prev;
                    if change_pct < -0.15 {
                        signals.push(format!("IV CRUSH: {:.1}% drop", change_pct * 100.0));
                    } else if change_pct > 0.20 {
                        signals.push(format!("IV SPIKE: {:.1}% increase", change_pct * 100.0));
                    }
                }
            }

            // Backwardation detection
            if let Some(s) = spread {
                if s > 0.05 {
                    signals.push(format!("BACKWARDATION: {:.1}%", s * 100.0));
                }
            }

            if !signals.is_empty() {
                found_signals = true;
                // Convert days since epoch to date
                if let Some(days) = date_days {
                    use chrono::{NaiveDate, Duration};
                    let epoch = NaiveDate::from_ymd_opt(1970, 1, 1).unwrap();
                    let date = epoch + Duration::days(days as i64);
                    println!("{}: {}", date, signals.join(", "));
                }
            }
        }

        if !found_signals {
            println!("No significant signals detected in this period");
        }
    }

    println!();

    Ok(())
}

File: scripts/view_atm_iv.py
#!/usr/bin/env python3
"""
Quick viewer for ATM IV parquet files
"""
import sys
import pyarrow.parquet as pq

def view_parquet(filepath):
    """Read and display parquet file contents"""
    table = pq.read_table(filepath)
    df = table.to_pandas()

    print("=" * 100)
    print(f"ATM IV Time Series: {filepath}")
    print("=" * 100)
    print()

    # Display all rows
    print("Data:")
    print("-" * 100)
    pd_opts = {
        'display.max_rows': None,
        'display.max_columns': None,
        'display.width': None,
        'display.max_colwidth': None,
        'display.float_format': lambda x: f'{x:.4f}' if pd.notna(x) else 'NaN'
    }

    import pandas as pd
    with pd.option_context(*[item for pair in pd_opts.items() for item in pair]):
        print(df.to_string(index=False))

    print()
    print("=" * 100)
    print("Summary Statistics:")
    print("=" * 100)
    print(df[['atm_iv_30d', 'atm_iv_60d', 'atm_iv_90d', 'term_spread_30_60', 'term_spread_30_90']].describe())
    print()

    # Detect potential earnings signals
    print("=" * 100)
    print("Potential Earnings Signals:")
    print("=" * 100)

    # Convert date function
    from datetime import datetime, timedelta
    epoch = datetime(1970, 1, 1)

    def date_num_to_str(date_num):
        """Convert days since epoch to YYYY-MM-DD string"""
        dt = epoch + timedelta(days=int(date_num))
        return dt.strftime('%Y-%m-%d')

    for i in range(1, len(df)):
        date = df.iloc[i]['date']
        prev_date = df.iloc[i-1]['date']

        iv_30_curr = df.iloc[i]['atm_iv_30d']
        iv_30_prev = df.iloc[i-1]['atm_iv_30d']
        term_spread = df.iloc[i]['term_spread_30_60']

        signals = []

        # IV Crush detection
        if pd.notna(iv_30_curr) and pd.notna(iv_30_prev) and iv_30_prev > 0:
            change_pct = (iv_30_curr - iv_30_prev) / iv_30_prev
            if change_pct < -0.15:  # 15% drop
                signals.append(f"IV CRUSH: {change_pct*100:.1f}% drop")
            elif change_pct > 0.20:  # 20% spike
                signals.append(f"IV SPIKE: {change_pct*100:.1f}% increase")

        # Backwardation detection
        if pd.notna(term_spread) and term_spread > 0.05:  # 5% backwardation
            signals.append(f"BACKWARDATION: {term_spread*100:.1f}%")

        if signals:
            date_str = date_num_to_str(date)
            print(f"{date_str}: {', '.join(signals)}")

    print()

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 view_atm_iv.py <parquet_file>")
        sys.exit(1)

    view_parquet(sys.argv[1])
